### DISCUSSION TITLE
GetExtension: vscode.typescript-language-features Not Found

### DISCUSSION THREAD

**Discussion by cemalgnlts** [CodinGame/monaco-vscode-api#346] [Q&A] [answered]
Hi,
I am trying to get the `typescript-language-features` package with `getExtension` but it returns `undefined`.

```js
console.log(api.extensions.allAcrossExtensionHosts);

// Result:
[
  { id: "vscode.css", ... },
  { id: "vscode.html", ... }
  { id: "vscode.typescript", ... },
  { id: "vscode.javascript", ... },
  { id: "vscode.typescript-language-features", ... },
];
```

```js
const tsExtension = api.extensions.getExtension("vscode.typescript-language-features");
console.log(tsExtension); // undefined
```

**Comment by cemalgnlts** [comment]
Fixed:

```diff
- const tsExtension = api.extensions.getExtension("vscode.typescript-language-features");
+ const tsExtension = api.extensions.getExtension("vscode.typescript-language-features", true);
```

But I didn't get the result I expected. I tried to imitate this extension:
https://github.com/esm-dev/esm.sh/blob/main/packages/esm-vscode/src/extension.ts#L172

I got this error:
```js
await tsExtension.activate(); // Cannot activate foreign extension.
```

**Reply by CGNonofr** [reply]
I don't think it's possible to make extensions running in a different extension host communicate

**Reply by cemalgnlts** [reply]
I think I explained it wrong. I'm trying something similar to this: https://code.visualstudio.com/api/references/vscode-api#extensions

I think this extension adds type files in this way. https://github.com/esm-dev/esm.sh/blob/main/packages/esm-vscode/src/extension.ts#L172

https://code.visualstudio.com/updates/v1_71#_typescript-server-plugins-on-web

**Reply by CGNonofr** [reply]
Calling getExtension(...).activate() IS making 2 extensions communicate. By default `vscode.typescript-language-features` runs in the Worker extension host while you are probably trying to access it from the main thread which is another extension host. Did I still misunderstand?

**Reply by cemalgnlts** [reply]
Yes, you got it right.

**Reply by cemalgnlts** [reply]
Yes, I can now access tsApi without getting an error. But the `init` function is not called.

<details>
<summary>Wrong use</summary>
```js
/* eslint-disable @typescript-eslint/no-var-requires */
// eslint-disable-next-line no-undef
const vscode = require("vscode");

/**
 *
 * @param {import("vscode").ExtensionContext} context VSCode api
 */
function activate(context) {
  context.subscriptions.push(
    vscode.commands.registerCommand("packageManager.startTypeTest", async () => {
      await ensureTsApi();
    })
  );
}

async function ensureTsApi() {
  const tsExtension = vscode.extensions.getExtension("vscode.typescript-language-features", true);

  /* ... */

  console.log("ensureTsApi"); // It works.
}

function init() {
  console.log("init"); // It won't work.
}

// eslint-disable-next-line no-undef
module.exports = {
  activate,
  default: init,
  __esModule: true
};
```
</details>

**Reply by CGNonofr** [reply]
why would it be called?

**Reply by cemalgnlts** [reply]
The VSCode documentation says:

> The browser entrypoint must be a JavaScript module that exports a default function that activates the plugin:

> export default function init(modules: { typescript: typeof import("typescript/lib/tsserverlibrary") }) {
  ...
};

**Reply by CGNonofr** [reply]
I think you are confusing VSCode extensions and typescript server plugins

**Reply by cemalgnlts** [reply]
Yes, you are right, the way I used it was wrong.

I found a log in the console:
```
Fetch API cannot load extension-file://jside.package-manager/node_modules/ts-pm-plugin/package.json.
URL scheme "extension-file" is not supported.
```

I then registered this JSON file.
```js
// index.ts:
registerFileUrl("node_modules/ts-pm-plugin/package.json", new URL("./package.json", import.meta.url).href);

// package.json:
{
  "name": "ts-pm-plugin",
  "version": "0.0.1",
  "browser": "ts-pm-plugin.js"
}
```

And I found another log.
```
Fetch API cannot load extension-file://jside.package-manager/node_modules/ts-pm-plugin/ts-pm-plugin.js.
URL scheme "extension-file" is not supported.
```

And I registered this file:
```js
// index.ts:
registerFileUrl("node_modules/ts-pm-plugin/ts-pm-plugin.js", new URL("./ts-pm-plugin.js", import.meta.url).href);

// ts-pm-plugin.js:
/** @param {{ typescript: import("typescript/lib/tsserverlibrary") }} modules */
function init(modules) {
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  const ts = modules.typescript;

  /** @param {ts.server.PluginCreateInfo} info */
  function create(info) {
    // Set up decorator object
    const proxy = Object.create(null);

    for (let k of Object.keys(info.languageService)) {
      const x = info.languageService[k];
      // @ts-expect-error - JS runtime trickery which is tricky to type tersely
      proxy[k] = (...args) => x.apply(info.languageService, args);
    }

    info.project.projectService.logger.info("Hello from TypeScript Server Plugin!");

    return proxy;
  }

  return { create };
}

// eslint-disable-next-line no-undef
module.exports = init;
```

Although it loads the plugin file, it does not call the `init` function.

**Reply by CGNonofr** [reply]
If it fails to load the script, I'm not surprised the function inside the script is not executed

Can you please provide a reproduction repo?

**Reply by cemalgnlts** [reply, marked as answer]
I started to prepare the sample repo and suddenly it occurred to me to enable TS Server Log like in the demo and examine the logs. I saw that the code gave an error in the logs, the `module` was not defined. After seeing this error, I converted it to ESM format:
```diff
- module.exports = init;
+ export default init;
```

Now my plugin is working **thanks a lot**.
